archive/tar_unix.go

```diff
@@ -124,7 +124,7 @@ func handleTarTypeBlockCharFifo(hdr *tar.Header, path string) error {
 
 func handleLChmod(hdr *tar.Header, path string, hdrInfo os.FileInfo) error {
 	if hdr.Typeflag == tar.TypeLink {
-		if fi, err := os.Lstat(hdr.Linkname); err == nil && (fi.Mode()&os.ModeSymlink == 0) {
+		if _, err := os.Stat(path); os.IsExist(err) {
```

Comment:
Can you clarify `os.IsExists(err)` part? When file exists, `os.Stat` returns `nil` error, right?

Reply:
If `os.Chmod()` changes the mode of the _target_, wouldn't this be the same as skipping the whole `if hdr.Typeflag ...`, and unconditionally doing;

```go
if err := os.Chmod(path, hdrInfo.Mode()); err != nil && !os.IsNotExist(err) {
	return err
}
```

Was the intent here to set the mode of the link itself (and not the target?)

Reply:
> Was the intent here to set the mode of the link itself (and not the target?)

What the original code tried to do is not clear (the commit message that added it does not give any reasoning). The original code checks that if tar headers say the file is hardlink and the target is not a symlink then `os.Chmod()`.

My patch follows that but adds a check that the file exists before `os.Chmod()` for it. AFAICS,

```
if err := os.Chmod(path, hdrInfo.Mode()); err != nil && !os.IsNotExist(err) {
	return err
}
```

would also work.

Reply:
Did some quick digging to find out what this code was intended to do. Haven't looked deeply yet, but looking for the history of it:

The code looks to be copied from https://github.com/moby/moby, where this function was extracted to a function (accounting for windows) in https://github.com/moby/moby/commit/8228ee4b0fe05b856ca84ef8ea84ba29484d9b9c.

The original code for this was added in https://github.com/moby/moby/commit/ab181ce55707de6f47d662dcdf6eab9c6c040906, which was this PR: https://github.com/moby/moby/pull/11099, and was created to fix https://github.com/moby/moby/issues/10937

Reply:
OK, so IIUC, the code was for situations where the hardlink is pointing to a symlink, which links to something that was not yet extracted from the tar archive. (which (again, IIUC) can cause issues in a situation where the hardlink/symlink is received before the final target is received from the tar stream?

I'm not too familiar with this stuff, but looks like we need to tread with care, and be sure this doesn't cause a regression.

_At least_ looks like;

- we need to add proper comments to this code to explain what it's doing
- if fixes are needed, please also open a pull request in the https://github.com/moby/moby repository to have the same fix there 🤗

Reply:
@vbatts @tonistiigi @dmcgowan you may be more familiar with this; PTAL

Reply:
@thaJeztah Thanks for taking the time with this!

> I'm not too familiar with this stuff, but looks like we need to tread with care, and be sure
> this doesn't cause a regression.

I totally agree. It starts with getting the understanding of what is needed and why ;)

Reply:
@thaJeztah btw, can you confirm the failure on your side?

```sudo ctr i pull docker.io/library/clearlinux:base```

Reply:
How come the change from `os.Lstat()` to `os.Stat()`? If you are concerned with `chmod`'ing things, if it is a symlink (_how_ this could even be the case, i'm not sure) then you'll need the information of the link itself. Stat will just give the information on the underlying inode which would be the hardlink you're targeting. Either _may_ need to be chmod'ed.
I think the behaviour you're seeking would not be to edit this case, but add another that checks for the presence of the underlying file.

Reply:
> I think the behaviour you're seeking would not be to edit this case, but add another that checks for the presence of the underlying file.

Yes, so as @thaJeztah also suggested, 

```
if err := os.Chmod(path, hdrInfo.Mode()); err != nil && !os.IsNotExist(err) {
	return err
}
```

Reply:
ah right

Reply:
but let me comment still on:

> How come the change from os.Lstat() to os.Stat()

`os.Lstat()` does not follow symlinks which is what `os.Chmod()` does so currently there's a race that a missing file can get unnoticed... that is why I suggested to move to `os.Stat()`.

anyway, I'll submit an update to have it as suggested above.